NumericOutcome.md

```diff
@@ -0,0 +1,542 @@
+      .zip(endDigits)
+      .takeWhile { case (startDigit, endDigit) => startDigit == endDigit }
+      .map(_._1)
+    
+    (prefixDigits, startDigits.drop(prefixDigits.length), endDigits.drop(prefixDigits.length))
+}
+```
+
+Now we need the algorithms for computing the groupings.
+
+```scala
+def frontGroupings(
+    digits: Vector[Int], // The unique digits of the range's start
+    base: Int): Vector[Vector[Int]] = {
+  val nonZeroDigits = digits.reverse.zipWithIndex.dropWhile(_._1 == 0) // Endpoint Optimization
+  
+  if (nonZeroDigits.isEmpty) { // All digits are 0
+      Vector(Vector(0))
+  } else {
+    val fromFront = nonZeroDigits.init.flatMap { // Note the flatMap collapses the rows of the grouping 
+      case (lastImportantDigit, unimportantDigits) =>
+        val fixedDigits = digits.dropRight(unimportantDigits + 1)
+        (lastImportantDigit + 1).until(base).map { lastDigit => // Note that this range excludes lastImportantDigit and base
+          fixedDigits :+ lastDigit
+        }
+    }
+
+    nonZeroStartDigits.map(_._1).reverse +: fromFront // Add Endpoint
+  }
+}
+
+def backGroupings(
+    digits: Vector[Int], // The unique digits of the range's end
+    base: Int): Vector[Vector[Int]] = {
+  val nonMaxDigits = digits.reverse.zipWithIndex.dropWhile(_._1 == base - 1) // Endpoint Optimization
+
+  if (nonMaxDigits.isEmpty) { // All digits are max
+    Vector(Vector(base - 1))
+  } else {
+    // Here we compute the back groupings in reverse so as to use the same iteration as in front groupings
+    val fromBack = nonMaxDigits.init.flatMap { // Note the flatMap collapses the rows of the grouping 
+      case (lastImportantDigit, unimportantDigits) =>
+        val fixedDigits = digits.dropRight(unimportantDigits + 1)
+        0.until(lastImportantDigit).reverse.toVector.map { // Note that this range excludes lastImportantDigit
+          lastDigit =>
+            fixedDigits :+ lastDigit
+        }
+    }
+
+    fromBack.reverse :+ nonMaxDigits.map(_._1).reverse // Add Endpoint
+  }
+}
+
+def middleGrouping(
+    firstDigitStart: Int, // The first unique digit of the range's start
+    firstDigitEnd: Int): Vector[Vector[Int]] = { // The first unique digit of the range's end
+  (firstDigitStart + 1).until(firstDigitEnd).toVector.map { firstDigit => // Note that this range excludes firstDigitEnd
+    Vector(firstDigit)
+  }
+}
+```
+
+Finally we are able to use all of these pieces to compress a range to an approximately minimal number of outcomes (by ignoring digits).
+
+```scala
+def groupByIgnoringDigits(start: Long, end: Long, base: Int, numDigits: Int): Vector[Vector[Int]] = {
+    val (prefixDigits, startDigits, endDigits) = separatePrefix(start, end, base, numDigits)
+    
+    if (start == end) { // Special Case: Range Length 1
+        Vector(prefixDigits)
+    } else if (startDigits.forall(_ == 0) && endDigits.forall(_ == base - 1)) { // Total Optimization
+        Vector(prefixDigits)
```

Comment:
I think this doesn't work if there is no common prefix.

Reply:
If there is no common prefix and start digits are all `0` and end digits are all `base - 1` then that is just the entire range in which case there's no point in having a DLC

Reply:
Good point, but if you go with computing positive and negative outcomes separately it might start to make sense.